freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/RepairInstancesService.java

```diff
@@ -255,4 +267,25 @@ public OperationStatus rebootInstances(String accountId, RebootInstancesRequest
         }
         return operationToOperationStatusConverter.convert(operation);
     }
+
+    public DescribeFreeIpaResponse rebuild(String accountId, RebuildRequest rebuildRequest) {
+        ValidationResult.ValidationResultBuilder validationResultBuilder = ValidationResult.builder();
+        Optional<Stack> stack =
+                stackService.findByCrnAndAccountIdWithListsEvenIfTerminated(rebuildRequest.getEnvironmentCrn(), accountId, rebuildRequest.getSourceCrn());
+        if (stack.isEmpty()) {
+            validationResultBuilder.error("Stack not found.");
+        }
+        Optional<Stack> nonTerminatedStack = stackService.findByEnvironmentCrnAndAccountId(rebuildRequest.getEnvironmentCrn(), accountId);
+        if (nonTerminatedStack.isPresent()) {
+            validationResultBuilder.error("There is a stack which hasn't been terminated.");
+        }
+        ValidationResult validationResult = validationResultBuilder.build();
+        if (validationResult.hasError()) {
+            LOGGER.debug("RebuildRequest has validation error(s): {}.", validationResult.getFormattedErrors());
+            throw new BadRequestException(validationResult.getFormattedErrors());
+        }
+        CreateFreeIpaRequest createFreeIpaRequest = stackToCreateFreeIpaRequestConverter.convert(stack.get());
```

Comment:
currently the `stack.get()` is not ensured explicitly by the code. I would remove `ValidationResultBuilder` and just throw the error at the right place

Reply:
I am a not sure that I understand what is meant by the comment:

> currently the stack.get() is not ensured explicitly by the code.


We have a check to ensure `stack.get()` is valid here:
```
        if (stack.isEmpty()) {
            validationResultBuilder.error("Stack not found.");
        }
```

Reply:
I mean idea highlights that it isn't checked. That's because the validationbuilder is used, but I don't see too much benefit of using the validationbuilder, instead of just throwing BadRequest at the respective points.

Reply:
ok, i updated it to just throw exceptions.